include/ruby/internal/has/builtin.h

```diff
@@ -46,7 +46,7 @@
 #  * __has_builtin  only  since  GCC  10.   This section  can  be  made  more
 #  * granular. */
 # /* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970 */
-# define RBIMPL_HAS_BUILTIN(_) (RBIMPL_HAS_BUILTIN_ ## _)
+# define RBIMPL_HAS_BUILTIN(_) defined(RBIMPL_HAS_BUILTIN_ ## _)
```

Comment:
@shyouhei Unfortunately this doesn't work and fails some CI jobs, e.g.
https://github.com/ruby/ruby/pull/4428/checks?check_run_id=2467096081
```
/home/runner/work/ruby/ruby/src/include/ruby/internal/assume.h:44:1: error: this use of "defined" may not be portable [-Werror=expansion-to-defined]
   44 | #if RBIMPL_HAS_BUILTIN(__builtin_unreachable)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is there maybe a way to ignore `expansion-to-defined` warnings while inside the RBIMPL_HAS_BUILTIN macro?

Another way might be to `#define RBIMPL_HAS_BUILTIN___builtin_assume 0` below, but that might be a non-trivial list.

Do you have any other idea how to fix this?

Reply:
> Is there maybe a way to ignore expansion-to-defined warnings while inside the RBIMPL_HAS_BUILTIN macro?

No... Sadly.  GCC cannot have `_Pragma` inside of `#if`.  That's the ultimate reason why we hesitated `-Wundef` before: https://github.com/ruby/ruby/blob/ruby_3_0/include/ruby/internal/token_paste.h#L35-L38

> Another way might be to add `#define RBIMPL_HAS_BUILTIN___builtin_assume 0` below, but there be many of those.

Yes this seems to be the "polite" way.  Note also that there are non-GCC non-Clang compilers that partially support __builtin_something.  Oracle Developer Studio is one of such compiler.  They use the "Take config.h definition when available" section below.  You would also like to consider modifying there.

> Do you have any other idea how to fix this?

- Dropping GCC <= 9 is one way.  Not practically possible today I guess, though.
- Write `#if (!defined(__GNUC__) && RBIMPL_HAS_BUILTIN(...))`  everywhere.